plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectCache.java

```diff
@@ -50,79 +52,103 @@ public OpenShiftProjectCache() {
 	}
 	
 	@Override
-	public synchronized Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
+	public Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
 		String key = getCacheKey(conn);
-		if(!cache.containsKey(key)) {
+		Set<IProjectAdapter> adapters = null;
+		if (!contains(key)) {
+			adapters = createEntry(key);
 			Collection<IProject> projects = conn.<IProject>getResources(ResourceKind.PROJECT);
-			Set<IProjectAdapter> adapters = new HashSet<>();
 			for (IProject project : projects) {
-				newProjectAdapter(adapters, conn, project);
+				IProjectAdapter adapter = addNewProjectAdapter(adapters, conn, project);
+				notifyAdd(adapter);
 			}
-			cache.put(key, adapters);
-			notifyAdd(adapters);
+		} else {
+			adapters = cache.get(key);
 		}
-		return new ArrayList<>(cache.get(key));
+
+		return new ArrayList<>(adapters);
+	}
+
+	private boolean contains(String key) {
+		return cache.containsKey(key);
 	}
-	
-	
 	
 	@Override
-	public synchronized void flushFor(IOpenShiftConnection conn) {
+	public void flushFor(IOpenShiftConnection conn) {
 		Set<IProjectAdapter> adapters = cache.remove(getCacheKey(conn));
-		if(adapters != null) {
-			notifyRemove(adapters);
+		if (adapters != null) {
+			synchronized(adapters) {
+				notifyRemove(adapters);
+			}
 		}
 	}
 
 	@SuppressWarnings("unchecked")
 	@Override
-	public synchronized void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
+	public void connectionChanged(IConnection connection, String property, Object oldValue, Object newValue) {
 		if (!(connection instanceof IOpenShiftConnection)) {
 			return;
 		}
-		if (ConnectionProperties.PROPERTY_PROJECTS.equals(property) && (oldValue instanceof List) && (newValue instanceof List)) {
-			List<IProject> newProjects = (List<IProject>) newValue;
-			List<IProject> oldProjects = (List<IProject>) oldValue;
-			IOpenShiftConnection conn = (IOpenShiftConnection)connection;
-			final String key = getCacheKey(conn);
-			
-			Set<IProjectAdapter> adapters = (Set<IProjectAdapter>) ObjectUtils.defaultIfNull(cache.get(key), new HashSet<>());
-			Map<IProject, IProjectAdapter> projectMap = adapters.stream().collect(Collectors.toMap(IProjectAdapter::getProject, Function.identity()));
-			
+		if (ConnectionProperties.PROPERTY_PROJECTS.equals(property) 
+				&& (oldValue instanceof List)
+				&& (newValue instanceof List)) {
+			IOpenShiftConnection openshiftConnection = (IOpenShiftConnection) connection;
+			String key = getCacheKey(openshiftConnection);
+			Set<IProjectAdapter> adapters = getOrCreateEntry(key);
+			addAndRemoveAdaptersFor((List<IProject>) newValue, (List<IProject>) oldValue, openshiftConnection, adapters);
+		}
+	}
+
+	private Set<IProjectAdapter> getOrCreateEntry(String key) {
+		Set<IProjectAdapter> adapters = cache.get(key);
```

Comment:
yes you're right. We discussed this and I then got the wrong assumption that the concurrent hashmap would guarantee this. But reading the javadoc clearly states that retrieval- and update-operations may overlaps since retrieval is not locking at all, only updates do.